sda_download/sda_download.go

```diff
@@ -0,0 +1,259 @@
+	var files []string
+	// Call ParseArgs to take care of all the flag parsing
+	err := helpers.ParseArgs(args, Args)
+	if err != nil {
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
+	}
+
+	if *datasetID == "" || *url == "" || *configPath == "" {
+		return fmt.Errorf("missing required arguments, dataset, config and url are required")
+	}
+
+	// Check that input file/folder list is not empty
+	if len(Args.Args()) == 0 {
+		return errors.New("no files to download")
+	}
+
+	files = append(files, Args.Args()...)
+
+	// Get the configuration file or the .sda-cli-session
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	// Check if the token has expired
+	err = helpers.CheckTokenExpiration(config.AccessToken)
+	if err != nil {
+		return err
+	}
+
+	// Loop through the files and download them
+	for _, filePath := range files {
+		downloadurl, err := downloadURL(*url, config.AccessToken, *datasetID, filePath)
+		if err != nil {
+			return err
+		}
+
+		filePathSplit := strings.Split(filePath, "/")
+		if strings.Contains(filePath, "elixir-europe.org") {
+			filePath = strings.Join(filePathSplit[1:], "/")
+		}
+
+		outFilename := filePath
+		if *outDir != "" {
+			outFilename = *outDir + "/" + filePath
+		}
+
+		err = downloadFile(downloadurl, config.AccessToken, outFilename)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// downloadFile downloads the file by using the download URL
+func downloadFile(uri, token, filePath string) error {
+	filePath = strings.TrimSuffix(filePath, ".c4gh")
+	// Get the file body
+	body, err := getResponseBody(uri, token)
+	if err != nil {
+		return fmt.Errorf("failed to get file for download, reason: %v", err)
+	}
+
+	// Create the directory if it does not exist
+	fileDir := filepath.Dir(filePath)
+	err = os.MkdirAll(fileDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create directory, reason: %v", err)
+	}
+
+	outfile, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to create file, reason: %v", err)
+	}
+	defer outfile.Close()
+
+	// Create a new progress container
+	p := mpb.New()
+
+	// Create a new progress bar with the length of the body
+	bar := p.AddBar(int64(len(body)),
+		mpb.PrependDecorators(
+			decor.CountersKibiByte("% .2f / % .2f"),
+		),
+	)
+
+	// Create a proxy reader
+	reader := strings.NewReader(string(body))
+	proxyReader := bar.ProxyReader(reader)
+
+	fmt.Printf("Downloading file to %s\n", filePath)
+	// Copy from the proxy reader (which updates the progress bar) to the file
+	_, err = io.Copy(outfile, proxyReader)
+	if err != nil {
+		return fmt.Errorf("failed to write file, reason: %v", err)
+	}
+
+	// Wait for the progress bar to finish
+	p.Wait()
+
+	return nil
+}
+
+// downloadURL gets the datset files, parses the JSON response to get the file ID
+// and returns the download URL for the file
+func downloadURL(baseURL, token, dataset, filename string) (string, error) {
```

Comment:
This should have a better name, something like `getFileIDURL`

Reply:
done